security/src/main/java/io/airlift/security/utils/KerberosUtil.java

```diff
@@ -0,0 +1,243 @@
+                    GSSManager gssManager = GSSManager.getInstance();
+
+                    if (initiator) {
+                        GSSName gssUserName = gssManager.createName(subjectPrincipal.getName(), GSSName.NT_USER_NAME);
+                        GSSCredential clientCredential = gssManager.createCredential(
+                                gssUserName,
+                                GSSCredential.DEFAULT_LIFETIME,
+                                KERBEROS_OID,
+                                GSSCredential.INITIATE_ONLY);
+
+                        GSSName gssServiceName = gssManager.createName(
+                                servicePrincipal,
+                                null);
+
+                        context = gssManager.createContext(
+                                gssServiceName,
+                                SPNEGO_OID,
+                                clientCredential,
+                                GSSContext.INDEFINITE_LIFETIME);
+                    }
+                    else {
+                        GSSName gssServiceName = gssManager.createName(servicePrincipal, null);
+                        GSSCredential serverCredential = gssManager.createCredential(
+                                gssServiceName,
+                                GSSCredential.INDEFINITE_LIFETIME,
+                                SUPPORTED_OIDS,
+                                GSSCredential.ACCEPT_ONLY);
+
+                        context = gssManager.createContext(serverCredential);
+                    }
+                    context.requestMutualAuth(true);
+                    context.requestConf(true);
+                    context.requestInteg(true);
+                    context.requestCredDeleg(false);
+
+                }
+                catch (GSSException e) {
+                    throw Throwables.propagate(e);
+                }
+                return context;
+            }
+        });
+        return gssContext;
+    }
+
+    public static class KerberosConfiguration
+            extends Configuration
+    {
+        public static final String GSS_SERVER = "gss-server";
+        public static final String GSS_CLIENT = "gss-client";
+        private final String principal;
+        private final String krb5Conf;
+        private static final Logger log = Logger.get(KerberosConfiguration.class);
+
+        public KerberosConfiguration(String subjectName, String krb5Config, boolean fromClient)
+                throws UnknownHostException
+        {
+            if (fromClient) {
+                this.principal = subjectName;
+            }
+            else {
+                this.principal = KerberosUtil.getServicePrincipal(subjectName, null);
+            }
+            this.krb5Conf = krb5Config;
+        }
+
+        @Override
+        public AppConfigurationEntry[] getAppConfigurationEntry(String name)
+        {
+            System.setProperty("java.security.krb5.conf", krb5Conf);
```

Comment:
Good point! We should only set it once when the Kerberos logic is first initialized both on the client and on the server-side.